Add AutoHighlight: Video highlights generation pipeline with Azure AI

kjindelmsft opened the pull request:
- Added complete video highlights generation pipeline
- Includes Jupyter notebook for easy execution
- Features Azure Content Understanding integration
- Supports schema-based video analysis
- Includes video stitching with moviepy
- Complete dependency management with requirements.txt

## Purpose
<!-- Describe the intention of the changes being proposed. What problem does it solve or functionality does it add? -->
* ...

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code
<!-- Add steps to run the tests suite and/or manually test -->
```
```

## What to Check
Verify that the following are valid
* ...

## Other Information
<!-- Add any other helpful information that may be needed here. -->

kjindelmsft committed:
Add AutoHighlight: Video highlights generation pipeline with Azure AI

- Added complete video highlights generation pipeline
- Includes Jupyter notebook for easy execution
- Features Azure Content Understanding integration
- Supports schema-based video analysis
- Includes video stitching with moviepy
- Complete dependency management with requirements.txt

kjindelmsft committed:
Fix AutoHighlight: Add actual files instead of submodule

gablans commented on AutoHighlight/build_highlight.py:
```diff
@@ -0,0 +1,250 @@
+PERSONALIZATION = DEFAULT_PERSONALIZATION
+OUT_DIR = "."
+EVENT_FIELD = "EventType"
+ENDPOINT = "https://ruibingptv2.openai.azure.com/"
```
Woah!  exposing credentials right here

gablans commented on AutoHighlight/build_highlight.py:
```diff
@@ -0,0 +1,250 @@
+ENDPOINT = "https://ruibingptv2.openai.azure.com/"
+DEPLOYMENT = "o1"
+API_VERSION = "2024-12-01-preview"
+API_KEY = "50d6d2ec54eb457da8a7bf2c35933ebf"
```
You are exposing API KEY

gablans reviewed (changes requested):
Left comments by line as well.

kjindelmsft committed:
Update AutoHighlight feature: add new files, update scripts, and remove obsolete files

kjindelmsft committed:
Fix AutoHighlight configuration flow: ensure pipeline configuration parameters are properly passed to highlight generation

yungshinlintw commented on AutoHighlight/README_NEW.md:
```diff
@@ -0,0 +1,184 @@
+# AutoHighlight: AI-Powered Video Highlights Generation
```
Please update the main README.md (https://github.com/Azure-Samples/azure-ai-content-understanding-with-azure-openai-python/blob/main/README.md) to point to this sample, plus a quick intro about AutoHighlight

yungshinlintw commented on AutoHighlight/build_highlight.py:
```diff
@@ -0,0 +1,276 @@
+import os, sys, json, math, time, pathlib, datetime, textwrap, re
```
nit: Python convention is to name module dir as 'auto_highlight"

yungshinlintw commented on AutoHighlight/README_NEW.md:
```diff
@@ -0,0 +1,184 @@
+# AutoHighlight: AI-Powered Video Highlights Generation
+
+Welcome! AutoHighlight is an innovative AI-powered pipeline that automatically generates compelling video highlights from any video content. By combining **Azure Content Understanding** with **OpenAI's advanced reasoning capabilities**, it intelligently identifies the most engaging moments and creates professional highlight reels tailored to your specific needs.
```
The file name should be just `README.md`. Currently, it's `README_NEW.md`

yungshinlintw commented on AutoHighlight/highlights_notebook.ipynb:
```diff
@@ -0,0 +1,508 @@
+    "import sys\n",
+    "import importlib.util\n",
+    "\n",
+    "# Complete package list for AutoHighlight\n",
```
Any reason that we just don't do a `%pip install -r ../requirements.txt`? Anything special that we are doing here?

yungshinlintw commented on AutoHighlight/run_video_analysis.py:
```diff
@@ -0,0 +1,255 @@
+            print(f"Details: {e}")
+            return None
+
+def create_or_update_analyzer(azure_ai_endpoint, analyzer_id, schema_path, access_token, api_version=None):
```
Please use the client in python/content_understanding_client.py to replace these code. You can see the use of the example in Gabe's PR: https://github.com/Azure-Samples/azure-ai-content-understanding-python/issues/77

yungshinlintw commented on AutoHighlight/json_parser.py:
```diff
@@ -0,0 +1,107 @@
+            return field_obj[key]
+    return None
+
+def extract_segments(analysis, segment_field="Segments"):
+    out = []
+    for content in analysis["result"]["contents"]:
+        fields = content.get("fields", {})
+        if segment_field not in fields:
+            continue
+        for seg_wrapper in fields[segment_field]["valueArray"]:
+            flat = {k: _extract_value(v) for k, v in seg_wrapper["valueObject"].items()}
+            out.append(flat)
+    out.sort(key=lambda s: s.get("StartTimeMs", 0))
+    return out
+
+def smart_highlight_filter(segments, k_start=4, k_end=4,
+                          event_field="PlayEvent", score_field="HighlightScore",
+                          rare_event_weight=1.5, min_per_event=1, n_quantile=0.4):
```
Please add function-level comments to explain these functions

yungshinlintw commented on AutoHighlight/video_analysis_schema.json:
```diff
@@ -0,0 +1,100 @@
+{
+  "description": "Eric and Asha Keynote AI Event highlight analyzer \u2013 segments video at key narrative moments and extracts flat metadata for each clip.",
```
Change the description given that this is a common schema?

yungshinlintw commented on AutoHighlight/build_highlight.py:
````diff
@@ -0,0 +1,276 @@
+    ```json
+    {draft}
+    ```""")
+    critique = safe_chat(client, [system, {"role":"user","content":checklist}],
+                         DEPLOYMENT)
+    if critique.strip() != "PASS":
+        revision_prompt = textwrap.dedent(f"""
+        Fix every issue below and return corrected JSON only.
+        Draft:
+        ```json
+        {draft}
+        ```
+        Issues:
+        ```json
+        {critique}
+        ```""")
+        final_txt = safe_chat(client, [system, {"role":"user","content":revision_prompt}],
+                              DEPLOYMENT)
+    else:
+        final_txt = draft
````
Consider Open AI Structured output to avoid the need to fix invalid json: 
https://platform.openai.com/docs/guides/structured-outputs

kjindelmsft committed:
Add AutoHighlight feature: AI-powered video highlights generation

- Reorganized files into Helper/ folder with CamelCase naming
- Created comprehensive HighlightsNotebook.ipynb pipeline
- Added modular Helper modules for video analysis and processing
- Implemented Azure Content Understanding + OpenAI integration
- Added professional README.md with full documentation
- Fixed path resolution issues for cross-platform compatibility
- Updated main README.md to include AutoHighlight sample

kjindelmsft committed:
Simplify package installation in HighlightsNotebook to use requirements.txt and clean up schema files

kjindelmsft committed:
Update README.md documentation

kjindelmsft committed:
Update RunVideoAnalysis to use Content Understanding Client and improve authentication handling

- Switch from custom HTTP requests to AzureContentUnderstandingClient
- Add dual authentication support (token + subscription key fallback)
- Fix 409 conflict error handling for existing analyzers
- Update API version to 2024-12-01-preview
- Add UUID-based unique analyzer ID generation
- Enhance error handling and logging

kjindelmsft committed:
Update BuildHighlight to use o3 model instead of o1

- Change deployment model from o1 to o3
- Update temperature handling to exclude o3 (doesn't support temperature)
- Increase max_completion_tokens to 50,000 for o3 compatibility
- Add clarifying comments for o3-specific configurations

kjindelmsft committed:
Update SchemaManager to use o3 model deployment and increase token limits

chienyuanchang commented:
@kjindelmsft thank you so much for your contribution! We’ve made some adjustments to integrate this project into the repository as a sample notebook, but we’ve kept the majority of your original designs and workflow in this PR: https://github.com/Azure-Samples/azure-ai-content-understanding-with-azure-openai-python/pull/16.
